mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -222,35 +222,32 @@ class CallPattern final : public SYCLToLLVMConversion<sycl::SYCLCallOp> {
   }
 
 private:
-  /// Rewrite sycl.call() {Function = *, Type = *} to a LLVM call to the
-  /// appropriate member function.
+  /// Rewrite sycl.call to a func call to the appropriate member function.
   LogicalResult rewriteCall(SYCLCallOp op, OpAdaptor opAdaptor,
                             ConversionPatternRewriter &rewriter) const {
     LLVM_DEBUG(llvm::dbgs() << "CallPattern: Rewriting op: "; op.dump();
                llvm::dbgs() << "\n");
 
     ModuleOp module = op.getOperation()->getParentOfType<ModuleOp>();
+    FuncBuilder builder(rewriter, op.getLoc());
 
-    Type retType = op.getODSResults(0).empty()
-                       ? LLVM::LLVMVoidType::get(module.getContext())
-                       : typeConverter.convertType(op.Result().getType());
+    bool producesResult = !op.getODSResults(0).empty();
+    TypeRange retType =
+        producesResult ? TypeRange(op.Result().getType()) : TypeRange();
+    func::CallOp funcCall =
+        builder.genCall(op.MangledName(), retType, op.getOperands(), module);
 
-    LLVMBuilder builder(rewriter, op.getLoc());
-    SmallVector<Type> operandTypes(opAdaptor.Args().getTypes());
-    FlatSymbolRefAttr funcRef = builder.getOrInsertFuncDecl(
-        opAdaptor.MangledName(), retType, operandTypes, module);
-    auto newOp = rewriter.replaceOpWithNewOp<LLVM::CallOp>(
-        op.getOperation(), op.getNumResults() == 0 ? TypeRange() : retType,
-        funcRef, opAdaptor.getOperands());
+    ValueRange results =
+        producesResult ? ValueRange(funcCall.getResult(0)) : ValueRange();
+    rewriter.replaceOp(op.getOperation(), results);
 
     LLVM_DEBUG({
       Operation *func = op->getParentOfType<LLVM::LLVMFuncOp>();
```

Comment:
Should we use `funcCall` instead of `op`, as it may not be valid after `replaceOp`?

Reply:
Changed